Add some doc regarding the new features in the maven plugin (#17547)
quarkusio/quarkus: docs/src/main/asciidoc/maven-tooling.adoc

More information on this topic can be found on the link:cdi-reference#bean_discovery[Bean Discovery] section of the CDI guide.

[[maven-configuration-profile]]
=== Building with a specific configuration profile

Quarkus supports xref:config-reference.adoc#profiles[configuration profiles] in order to provide a specific configuration according to the target environment.

The profile can be provided directly in the maven build's command thanks to the System property `quarkus.profile` with a command of type `./mvnw package -Dquarkus.profile=<profile-name-here>`,
however it is also possible to specify the profile directly into the pom file of your project either using the properties of project, the system properties of the plugin or the properties of the plugin.

In order of precedence (greater precedence first):

> System properties of the plugin:

Reviewer: ```suggestion
1. System properties set in the Quarkus Maven plugin configuration
```

[source,xml]
----
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  ...
  <build>
    <plugins>
      ...
      <plugin>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <version>${quarkus-plugin.version}</version>
        <extensions>true</extensions>
        <configuration>
          <systemProperties>
            <quarkus.profile>prod-aws</quarkus.profile> <1>
          </systemProperties>
        </configuration>
     </plugin>
     ...
    </plugins>
  </build>
...
</project>
----
<1> The default configuration profile of this project is `prod-aws`.

> Properties of the plugin:

Reviewer: ```suggestion
2. Quarkus Maven plugin configuration properties
```

[source,xml]
----
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  ...
  <build>
    <plugins>
      ...
      <plugin>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-maven-plugin</artifactId>
        <version>${quarkus-plugin.version}</version>
        <extensions>true</extensions>
        <configuration>
          <properties>
            <quarkus.profile>prod-aws</quarkus.profile> <1>
          </properties>
        </configuration>
     </plugin>
     ...
    </plugins>
  </build>
...
</project>
----
<1> The default configuration profile of this project is `prod-aws`.

> Properties of the project:

Reviewer: ```suggestion
3. Project properties
```

[source,xml]
----
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  ...
  <properties>
    <quarkus.profile>prod-aws</quarkus.profile> <1>
    ...
  </properties>
...
</project>
----
<1> The default configuration profile of this project is `prod-aws`.

> NOTE: Whatever the approach chosen, it can still be overridden with the System property `quarkus.profile` or the environment variable `QUARKUS_PROFILE`.

Reviewer: ```suggestion
NOTE: Whatever the approach is chosen, the profile can still be overridden with the `quarkus.profile` system property or the `QUARKUS_PROFILE` environment variable.
```

[[maven-multi-build]]
=== Building several artifacts from a single module

In some particular use cases, it can be interesting to build several artifacts of your application from the same module.
A typical example is when you want to build your application with different configuration profiles.

> In that case, it is possible to add as many executions as needed to the declaration of the Quarkus maven plugin like in the following example.

Reviewer: ```suggestion
In that case, it is possible to add as many executions as needed to the Quarkus Maven plugin configuration.
```

> Example where the plugin will build the application twice one for Oracle and the other for PostgreSQL:

Reviewer: ```suggestion
Below is an example of a Quarkus Maven plugin configuration that will produce two builds of the same application: one  using the `prod-oracle` profile and the other one using the `prod-postgresql` profile.
```

              </properties>
            </configuration>
          </execution>
          <execution>
            <id>postgresql</id>
            <goals>
              <goal>build</goal>
            </goals>
            <configuration>
              <properties>
                <quarkus.profile>prod-postgresql</quarkus.profile> <3>
                <quarkus.package.output-directory>postgresql-quarkus-app</quarkus.package.output-directory> <4>
              </properties>
            </configuration>
          </execution>
        </executions>
     </plugin>
     ...
    </plugins>
  </build>
...
</project>
----
<1> The default configuration profile of the first execution of the plugin is `prod-oracle`.
<2> The output directory of the first execution of the plugin is set to `oracle-quarkus-app` instead of `quarkus-app` to have a dedicated directory.
<3> The default configuration profile of the second execution of the plugin is `prod-postgresql`.
<4> The output directory of the second execution of the plugin is set to `postgresql-quarkus-app` instead of `quarkus-app` to have a dedicated directory.

> Since the plugin builds the application with the same dependencies, we may end up with dependencies that are not needed according to the configuration profile used.

Reviewer: ```suggestion
NOTE: With the configuration above, both profile builds will be using the same dependencies, so if we added dependencies on the Oracle and PostgreSQL drivers to the application, both of the drivers will appear in both builds.
```

> Indeed, in our previous example, we would like to have the JDBC driver for Oracle only in the final artifact generated with the configuration profile `prod-oracle` and the same thing for PostgreSQL.

Reviewer: ```suggestion
```

> By default, the final artifacts will contain both divers which is not what we want, to avoid that you can define both JDBC drivers as optional dependencies then indicate Quarkus that the optional dependencies need to be filtered and how as next:

Reviewer: ```suggestion
To isolate profile-specific dependencies from other profiles, the JDBC drivers could be added as optional dependencies to the application but configured to be included in each profile that requires them, e.g.:
```
